contrib/diurnal/dc.go

```diff
@@ -0,0 +1,293 @@
+					break
+				}
+				c := smoothed[len(smoothed)-1].count + sign*int(maxRate)
+				t := smoothed[len(smoothed)-1].time
+				if sign*c >= sign*tc[i].count {
+					continue // don't need this, will exit on following iteration
+				}
+				smoothed = append(smoothed, timeCount{t + time.Minute, c})
+			}
+		}
+		smoothed = append(smoothed, tc[i])
+	}
+	return smoothed, nil
+}
+
+var posError = errors.New("could not find position")
+
+func findPos(tc []timeCount, cur int, offset time.Duration) int {
+	first := true
+	for i := cur; i != cur || first; i = (i + 1) % len(tc) {
+		if tc[i].time > offset {
+			return i
+		}
+		first = false
+	}
+	return 0
+}
+
+func (s *Scaler) setCount(c int) {
+	glog.Infof("scaling to %d replicas", c)
+	rcList, err := client.ReplicationControllers(*namespace).List(s.selector)
+	if err != nil {
+		glog.Errorf("could not get replication controllers: %v", err)
+		return
+	}
+	for _, rc := range rcList.Items {
+		rc.Spec.Replicas = c
+		if _, err = client.ReplicationControllers(*namespace).Update(&rc); err != nil {
+			glog.Errorf("unable to scale replication controller: %v", err)
+		}
+	}
+}
+
+func (s *Scaler) timeOffset() time.Duration {
+	return time.Since(s.start) % (24 * time.Hour)
+}
+
+func (s *Scaler) curpos(offset time.Duration) int {
+	return findPos(s.timeCounts, s.pos, offset)
+}
+
+func (s *Scaler) scale() {
+	for {
+		select {
+		case <-s.done:
+			return
+		default:
+			offset := s.timeOffset()
+			s.pos = s.curpos(offset)
+			if s.timeCounts[s.pos].time < offset {
+				time.Sleep(24*time.Hour - offset)
+				continue
+			}
+			time.Sleep(s.timeCounts[s.pos].time - offset)
+			s.setCount(s.timeCounts[s.pos].count)
+		}
+	}
+}
+
+func (s *Scaler) Start() error {
+	now := time.Now().UTC()
+	s.start = time.Date(now.Year(), now.Month(), now.Day(), 0, 0, 0, 0, now.Location())
+	if *startNow {
+		s.start = now
+	}
+
+	// set initial count
+	pos := s.curpos(s.timeOffset())
+	// add the len to avoid getting a negative index
+	pos = (pos - 1 + len(s.timeCounts)) % len(s.timeCounts)
+	s.setCount(s.timeCounts[pos].count)
+
+	s.done = make(chan struct{})
+	go s.scale()
+	return nil
+}
+
+func safeclose(c chan<- struct{}) (err error) {
+	defer func() {
+		if e := recover(); e != nil {
+			err = e.(error)
+		}
+	}()
+	close(c)
+	return nil
+}
+
+func (s *Scaler) Stop() error {
+	if err := safeclose(s.done); err != nil {
+		return errors.New("already stopped scaling")
+	}
+	return nil
+}
+
+func parseLabelsSelector(csvLabelSelector string) (labels.Selector, error) {
+	var s labels.Selector = labels.LabelSelector{}
+	strSelector := strings.Split(csvLabelSelector, ",")
+	for i := range strSelector {
+		kv := strings.Split(strSelector[i], "=")
+		if len(kv) != 2 {
+			return nil, errors.New("invalid label selector, must be of form key=value")
+		}
+		s = s.Add(kv[0], "=", kv[1:])
+	}
+	return s, nil
+}
+
+var (
+	counts       = flag.String("counts", "1", "replica counts, must have at least one (csv)")
```

Comment:
It looks like the assumption is that counts and times must be the same len(), but that is not called out.